### PR TITLE
InvocationMonitor fix problem with task failure

### PR DESCRIPTION
**Pull request by pveentjer**
Fixes a problem in the InvocationMonitor in case of task error. If a task executed by the scheduler throws an exception, and this task is a periodic executed task like scanning invocations, it will
not be executed again. This is solved by putting a try/catch around all tasks and dropping the uncaughtexceptionhandler from the thread.

Also some logging has been added to see if periodic tasks like scanning invocations is delayed too much. Also a probe has been added to record the number of delays.

### CODE REVIEW

**Comment by pveentjer** [comment]
I'm still not terribly excited about logging the time. I'm thinking about being able to calculate the next running time of certain tasks and if there is too much of a delay with the expected time and actual time, the task can do some logging.

**Comment by pveentjer** [comment]
The logging has been improved so that a fixed rate task will show if it has been delayed.

**Comment by pveentjer** [comment]
run-lab-run

**Comment by jerrinot** [comment]
:+1:

File: hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvocationMonitor.java [hunk 1]
```diff
@@ -212,6 +214,50 @@ long getLastMemberHeartbeatMillis(Address memberAddress) {
         return heartbeat == null ? 0 : heartbeat.get();
     }
 
+    private abstract class MonitorTask implements Runnable {
+        @Override
+        public void run() {
+            try {
+                run0();
+            } catch (Throwable t) {
+                inspectOutputMemoryError(t);
```
Comment (mtopolnik) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvocationMonitor.java]: Maybe this is a good opportunity to fix the name of this method? `inspectOutOfMemoryError`.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvocationMonitor.java [hunk 2]
```diff
@@ -212,6 +214,50 @@ long getLastMemberHeartbeatMillis(Address memberAddress) {
...
+            this.periodMillis = periodMillis;
+        }
+
+        @Override
+        public void run() {
+            long currentTimeMillis = System.currentTimeMillis();
+
+            try {
+                if (expectedNextMillis + MAX_DELAY_MILLIS < currentTimeMillis) {
+                    logger.warning(getClass().getSimpleName() + " delayed " + (currentTimeMillis - expectedNextMillis) + " ms");
```
Comment (mtopolnik) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvocationMonitor.java]: Would it contribute anything to triage the delay causes into "last task execution took too long" and "next task failed to be scheduled on time"?

**Comment by mtopolnik** [comment]
👍 (minor comments only)

File: hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvocationMonitor.java (hunk 1 above)
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvocationMonitor.java]: What do you suggest?

Comment (mtopolnik) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvocationMonitor.java]: I already suggested :)

Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvocationMonitor.java]: Ahhh.. I see it.

**Pull request merged**